### ISSUE TITLE
Reconnect missing when broker closes connection gracefully

### ISSUE THREAD

**Issue by mxscho** [benbria/node-amqp-connection-manager#8]
Hi,

My CoffeeScript skills are not exactly brilliant - but after investigating [this part of the code](https://github.com/benbria/node-amqp-connection-manager/blob/master/src/AmqpConnectionManager.coffee#L104) I would guess that a reconnection attempt is only initiated if the broker went away due to an unexpected error.
However, besides the `error` event there's also a `close` event which can be fired from the connection. This happens if you e.g. manually shutdown the RabbitMQ server. The underlying amqplib then generates a `CONNECTION_FORCED` error and emits a `close` event for the connection.

As far as I can tell there just needs to be another event handler for connections' `close` events implemented which also tries to do a reconnect. However as mentioned, I haven't been doing anything in CoffeeScript yet and that's why I didn't want to do a pull request right away.

EDIT: I just saw the (last) test which makes sure that there's no reconnection attempt after a connection close. So furthermore there should be a distinction between close events fired by us manually closing the connection in the client and those fired by the server as described in the above example.

Feel free to correct me if I'm wrong.

**Comment by jwalton** [comment]
Hmm...  It's weird I haven't caught this before, because we had a problem where we were not sending heartbeats to the RabbitMQ server for a short while, and RabbitMQ kills the connection, but we seem to reconnect.  I'll have a look at this, for sure.

**Comment by jwalton** [comment]
So, I'm on MacOS.  I wrote a little test app here:

```
amqp = require('.');

var handleMessage = function() {
    console.log("message!");
};

// Create a new connection manager
var connection = amqp.connect(['amqp://localhost'], {json: true});

connection.on('disconnect', (opts) => console.log("disconnect!", opts.err ? opts.err.stack : ''));
connection.on('connect', () => console.log("connect!"));
```

Then I shut down RabbitMQ with:

    launchctl unload ~/Library/LaunchAgents/homebrew.mxcl.rabbitmq.plist

This gets me:

    disconnect! Error: Unexpected close

and then more disconnect messages as we try to reconnect (which is weird, but it's how I wrote it. :P)  If I restart RabbitMQ, then it reconnects correctly.

So, how are you recreating this?

I'm certainly open to adding the listener for "close", it would just be nice to figure out a way to test this and make sure I got it right.  :)

**Comment by mxscho** [comment]
Thanks for looking into this.

I would assume that `launchctl` probably just killed the RabbitMQ server process in your case so there's nothing the server can send to the client.
I'm working with RabbitMQ 3.3.5 on Debian 8 (Jessie).
If I do a

    service rabbitmq-server stop

[this method](https://github.com/squaremo/amqp.node/blob/master/lib/connection.js#L76) in amqplib gets called.
I don't know anything about the AMQP, but I think that there's something like a close packet a server can send to the client to tell the client that it's going to shut down.

What's happening now is that this method [checks if the error occurred is fatal](https://github.com/squaremo/amqp.node/blob/master/lib/connection.js#L92). In this case the generated error contains the code `defs.constants.CONNECTION_FORCED` which I think represents a closed connection initiated by the server.

Due to the fact that it's not a fatal error, the connection does not emit an `error` event and the amqp-connection-manager does not see this loss of the connection. Instead, [the function which is called at the end](https://github.com/squaremo/amqp.node/blob/master/lib/connection.js#L95) just emits a `close` event on the connection which is currently not handled by amqp-connection-manager.
As far as I understood - in e.g. our case of a graceful shutdown only a `close` event is emitted, whereas in case of an error both, an `error` and a `close` event are fired. Not quite sure though because I haven't been testing this.

If I do the above described service stop command, the `maybeErr` taken by [the `toClosed(...)` function](https://github.com/squaremo/amqp.node/blob/master/lib/connection.js#L398) contains following message:

    Connection closed: 320 (CONNECTION-FORCED) with message "CONNECTION_FORCED - broker forced connection closure with reason 'shutdown'"

I don't know if there are any other reasons besides `shutdown` right now, but I think this should definitely issue a reconnect in amqp-connection-manager.

If you need any more information or reproduction steps/descriptions feel free to ask me about it.

**Comment by jacky1967** [comment]
Lo,
Just add this : 
            connection.on('close', function () {
              return wait(_this.reconnectTimeInSeconds * 1000).then(function() {
                return _this._connect();
              });
            });
In the file : AmqpConnectionManager.js at line 117
After that everything work fine ;-)

**Comment by oceddi** [comment]
Any progress on this issue?  I'm also noticing missed reconnects on graceful close.

**Comment by jwalton** [comment]
I'm kinda swamped right now, but if someone wants to build a PR with @jacky1967 's change, I'll pull it in and publish a new version.

**Comment by oceddi** [comment]
I made a PR to fix this.

**Comment by jwalton** [comment]
Published as v1.3.2.  Thanks @oceddi!  Sorry it took so long to get this fixed.

**Comment by jwalton** [comment]
Hmm...  If we have an error in a channel, and it isn't caught by a channel error handler, then it bubbles up to the (not closed) connection, and we end up running the `connection.on("error")` handler followed by the `connection.on("close")` handler, and then we have two connections.  :(

**Comment by jwalton** [comment]
Fixed in v1.3.4.
